Added python3 compatability

cmonr opened the pull request:
A continuation of https://github.com/ARMmbed/mbed-cli/pull/609, and then some.

Tested within Linux, using two virtual environments, one being Py2 and the other being Py3.
Passes all py.test tests in both versions.

At the moment, this will not complete `python3 circle_tests.py` because there's at least one fix targeting 5.9.0-rc1 in mbed-os (https://github.com/ARMmbed/mbed-os/pull/6592). CircleCI uses the `latest` tag to run its tests, which means it won't pick up fixes in mbed-os' master branch.


CircleCI checks will need to be updated before this can pass CI tests and be brought in, but that shouldn't stop this from being reviewed.

cmonr commented:
^^^ Gotta love messy rebases.

cmonr commented on mbed/mbed.py:
```diff
@@ -1948,12 +1957,13 @@ def import_(url, path=None, ignore=False, depth=None, protocol=None, insecure=Fa
         if ignore:
             warning(err)
         else:
+            print("THERE")
```
xD

Should have just stuck with "HI" everywhere.

cmonr commented on test/util.py:
```diff
@@ -168,7 +170,7 @@ def testrepos(mbed, request):
                 elif scm() == 'hg':
                     hash = pquery(['hg', 'id', '-i'])
-            f.write(test4 + '/#' + hash + '\n')
+            f.write(test4 + '/#' + str(hash) + '\n')
```
Nice catch. The decode, and even the str() is no longer needed because the pquery function has been updated to do the decode automatically.

cmonr commented on mbed/mbed.py:
```diff
@@ -2412,8 +2422,8 @@ def compile_(toolchain=None, target=None, profile=False, compile_library=False,
               + ['-t', tchain, '-m', target]
-              + list(chain.from_iterable(izip(repeat('--profile'), profile or [])))
-              + list(chain.from_iterable(izip(repeat('--source'), source)))
+              + list(chain.from_iterable(list(zip(repeat('--profile'), profile or []))))
```
Eliminated the list by importing something similar.

cmonr commented:
Fyi, this depends on https://github.com/ARMmbed/mbed-cli/pull/668

cmonr commented:
If it's better to simply bring https://github.com/ARMmbed/mbed-cli/pull/668 into this, just lmk.

screamerbg commented on mbed/mbed.py:
```diff
@@ -251,12 +260,12 @@ def pquery(command, output_callback=None, stdin=None, **kwargs):
 
     if very_verbose:
-        log(str(stdout).strip()+"\n")
+      log(stdout.decode("utf-8").strip() + "\n")
```
Looks like the indent is lost (2 spaces instead of 4)

screamerbg commented:
@cmonr #668 is now merged to `master`. Could you please rebase?

cmonr commented:
@screamerbg Rebased and updated. 

Requesting re-review from @bridadan and @theotherjimmy.

Removed the hg tests from py3 for now. It should be put back in, but there's a decision that needs to be made about what the hg repos should be.

cmonr commented:
Otherwise, this is what the py3 hg failure looks like: https://circleci.com/gh/cmonr/mbed-cli/170
